### PR TITLE
Geometry View

### PR DESCRIPTION
**Pull request by antonmic**
## What does this PR do?

Introduced a new class called GeometryView, which wraps IndexBufferView, StreamBufferViews and DrawArguments. This has three benefits:
1. It de-duplicates data between drawItems and drawPackets, as these now hold a pointer to the respective GeometryView, making both DrawPackets and DrawItems smaller in size and thus more cache friendly
2. Avoid having to copy and recreate buffer views whenever a DrawPacket is recreated
3. Most importantly, it allows index and stream buffers and buffer views to be modified at runtime without having to rebuild DrawPackets/DrawItems. This allows us to change the underlying geometry at runtime on a per-frame basis without incurring the cost of per-frame DrawPacket construction.

Notes: DrawArguments now lives in it's own file DrawArguments.h (and DeviceDrawArguments.h)

Recommended to review these files first:
DeviceGeometryView/GeometryView
DeviceDrawItem/DrawItem
DeviceDrawPacket/DrawPacket
DeviceDrawPacketBuilder/DrawPacketBuilder
CommandList

## How was this PR tested?

Tested the following in Editor to ensure they render correctly:
Meshes
Skinned and Animated Actor (Rin)
Stars
Terrain
ImGui
DynamicDrawContext
AuxGeom
FixedShape
ReflectionProbe
GI Probe with live updates

Not tested:
Hair (please provide test instructions if you have any)
Multi-GPU rendering (please add any relevant multi-GPU reviewers to this review)

### CODE REVIEW

**Review by jhmueller-huawei — COMMENTED** [review]
I've had a look at the list of files you recommended to review first (not the others). So far looks good to me from a multi-device point of view. I haven't tried the code though. Do you have an accompanying atom-sampleviewer PR for these changes? To test multi-devices you can try the RPI -> MultiGPU sample there.

File: Gems/Atom/RHI/Code/Include/Atom/RHI/DeviceGeometryView.h [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+
+        bool operator==(const StreamBufferIndices& other) const
+        {
+            bool same = (m_count == other.m_count);
+            u32 arraySize = RHI::Limits::Pipeline::StreamCountMax / 2;
+            for (u32 i = 0; i < arraySize; ++i)
+            {
+                same = same && (m_indices[i] == other.m_indices[i]);
+            }
+            return same;
```
[the remark this answers is not in the document]
Comment (antonmic) [reply Gems/Atom/RHI/Code/Include/Atom/RHI/DeviceGeometryView.h:68]: Thanks, very good catch, I'll change the loop to use m_count.

File: Gems/Atom/RHI/Code/Include/Atom/RHI/GeometryView.h [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+                if (m_indexBufferView.IsValid())
+                {
+                    newGeometryView.SetIndexBufferView(m_indexBufferView.GetDeviceIndexBufferView(deviceIndex));
+                }
+                for (StreamBufferView& stream : m_streamBufferViews)
+                {
+                    newGeometryView.AddStreamBufferView(stream.GetDeviceStreamBufferView(deviceIndex));
+                }
+                newGeometryView.m_dummyStreamBufferIndex = m_dummyStreamBufferIndex;
+                m_geometryViews.emplace(deviceIndex, newGeometryView);
```
[the remark this answers is not in the document]
Comment (antonmic) [reply Gems/Atom/RHI/Code/Include/Atom/RHI/GeometryView.h:55]: Thanks, will fix

File: Gems/Atom/RHI/Code/Include/Atom/RHI/DrawPacketBuilder.h [hunk 3]
```diff
@@ -21,6 +22,10 @@ namespace AZ
 
 namespace AZ::RHI
 {
+    //! DrawPachetBuilder builds a DrawPacket and its DrawItems contiguously in memory for cache efficiency
```
Comment (galibzon) [inline Gems/Atom/RHI/Code/Include/Atom/RHI/DrawPacketBuilder.h:25]: Typo: `DrawPachetBuilder` should be `DrawPacketBuilder`

**Review by galibzon — APPROVED** [review]

File: Gems/Atom/RHI/Code/Source/RHI/GeometryView.cpp [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+#include <Atom/RHI/GeometryView.h>
+#include <Atom/RHI.Reflect/InputStreamLayout.h>
+
+namespace AZ::RHI
+{
+
+    bool ValidateStreamBufferViews(const InputStreamLayout& inputStreamLayout, RHI::GeometryView& geometryView,
+        const RHI::StreamBufferIndices& streamIndices)
+    {
+        bool ok = true;
```
Comment (moudgils) [inline Gems/Atom/RHI/Code/Source/RHI/GeometryView.cpp:18]: Minor recommendation. Rename ok -> validStreams

File: Gems/Atom/RHI/Code/Source/RHI/GeometryView.cpp [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+                }
+            }
+
+            if (streamCount != inputStreamLayout.GetStreamBuffers().size())
+            {
+                AZ_Assert(false, "InputStreamLayout references %d stream buffers but %d StreamBufferViews passed the mask check.",
+                    inputStreamLayout.GetStreamBuffers().size(), streamCount);
+                ok = false;
+            }
+
```
Comment (moudgils) [inline Gems/Atom/RHI/Code/Source/RHI/GeometryView.cpp:57]: remove empty line

File: Gems/Atom/RHI/Code/Tests/DrawPacketTests.cpp [hunk 6]
```diff
@@ -25,23 +25,19 @@ namespace UnitTest
...
+            m_streamIndices = geometryView.GetFullStreamBufferIndices();    // Ordered Stream Indices
 
             m_tag = RHI::DrawListTag(random.GetRandom() % RHI::Limits::Pipeline::DrawListTagCountMax);
             m_stencilRef = static_cast<uint8_t>(random.GetRandom());
             m_sortKey = random.GetRandom();
         }
 
-        AZStd::array<RHI::DeviceStreamBufferView, RHI::Limits::Pipeline::StreamCountMax> m_streamBufferViews;
-
+        const RHI::DeviceGeometryView* m_geometryView;
```
Comment (moudgils) [inline Gems/Atom/RHI/Code/Tests/DrawPacketTests.cpp:39]:
```suggestion
        const RHI::DeviceGeometryView* m_geometryView = nullptr;
```

File: Gems/Atom/RPI/Code/Include/Atom/RPI.Public/Pass/FullscreenTrianglePass.h [hunk 7]
```diff
@@ -83,6 +83,8 @@ namespace AZ
             // The draw item submitted by this pass
             RHI::DrawItem m_item;
 
+            RHI::GeometryView m_geometryView;
```
Comment (moudgils) [inline Gems/Atom/RPI/Code/Include/Atom/RPI.Public/Pass/FullscreenTrianglePass.h:86]: Add a quick comment to follow existing style

File: Gems/Atom/RPI/Code/Include/Atom/RPI.Public/Pass/Specific/ImageAttachmentPreviewPass.h [hunk 8]
```diff
@@ -136,6 +136,8 @@ namespace AZ
                 // The draw item for drawing the image preview for this type of image
                 RHI::DrawItem m_item{RHI::MultiDevice::AllDevices};
 
+                RHI::GeometryView m_geometryView;
```
Comment (moudgils) [inline Gems/Atom/RPI/Code/Include/Atom/RPI.Public/Pass/Specific/ImageAttachmentPreviewPass.h:139]: Add a quick comment to follow existing style

File: Gems/Atom/RHI/Code/Include/Atom/RHI/DeviceGeometryView.h [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+    //! To iterate through stream buffer views with these indices, use the StreamIterator class below
+    class StreamBufferIndices
+    {
+        u8 m_count = 0;
+        u8 m_indices[RHI::Limits::Pipeline::StreamCountMax / 2];
+
+    public:
+        void AddIndex(u8 index)
+        {
+            AZ_Assert(m_count < RHI::Limits::Pipeline::StreamCountMax, "Adding %d stream buffer indices, but the max count only allows for %d",
```
[the remark this answers is not in the document]
Comment (antonmic) [reply Gems/Atom/RHI/Code/Include/Atom/RHI/DeviceGeometryView.h:32]: Will fix

File: Gems/Atom/Feature/Common/Code/Source/AuxGeom/FixedShapeProcessor.cpp [hunk 10]
```diff
@@ -1357,14 +1357,21 @@ namespace AZ
...
 
-            objectBuffers.m_streamBufferViews = { positionBufferView };
-            objectBuffers.m_streamBufferViewsWithNormals = { positionBufferView, normalBufferView };
+            objectBuffers.m_pointGeometryView.AddStreamBufferView(positionBufferView);
+            objectBuffers.m_lineGeometryView.AddStreamBufferView(positionBufferView);
+            objectBuffers.m_triangleGeometryView.AddStreamBufferView(positionBufferView);
+            objectBuffers.m_triangleGeometryView.AddStreamBufferView(normalBufferView);
+
+            RHI::StreamBufferIndices streamIndices;
+            streamIndices.AddIndex(0);  // Positions
```
[the remark this answers is not in the document]
Comment (antonmic) [reply Gems/Atom/Feature/Common/Code/Source/AuxGeom/FixedShapeProcessor.cpp:1366]: Will fix

File: Gems/Atom/RHI/Code/Include/Atom/RHI/DrawArguments.h [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+            default:
+                return DeviceDrawArguments();
+            }
+        }
+
+        DrawType m_type;
+        union {
+            DrawIndexed m_indexed;
+            DrawLinear m_linear;
+            DrawIndirect m_mdIndirect;
```
[the remark this answers is not in the document]
Comment (antonmic) [reply Gems/Atom/RHI/Code/Include/Atom/RHI/DrawArguments.h:67]: Will fix

File: Gems/Atom/RHI/Code/Source/RHI/DrawPacketBuilder.cpp [hunk 12]
```diff
@@ -41,17 +31,7 @@ namespace AZ::RHI
 
     DrawPacketBuilder::DrawPacketBuilder(const DrawPacketBuilder& other)
     {
-        m_deviceMask = other.m_deviceMask;
-
-        m_drawRequests = other.m_drawRequests;
-
-        m_drawPacketInFlight = aznew DrawPacket;
```
[the remark this answers is not in the document]
Comment (antonmic) [reply Gems/Atom/RHI/Code/Source/RHI/DrawPacketBuilder.cpp:48]: Should be?
You mean a shared pointer?

File: Gems/Atom/RHI/Code/Source/RHI/GeometryView.cpp (hunk 4 above)
Comment (antonmic) [reply Gems/Atom/RHI/Code/Source/RHI/GeometryView.cpp:18]: Will do

File: Gems/Atom/RHI/Code/Tests/DrawPacketTests.cpp (hunk 6 above)
Comment (antonmic) [reply Gems/Atom/RHI/Code/Tests/DrawPacketTests.cpp:39]: Will fix

File: Gems/Atom/RHI/Code/Source/RHI/DrawPacketBuilder.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (moudgils) [reply Gems/Atom/RHI/Code/Source/RHI/DrawPacketBuilder.cpp:48]: I meant to say that 'this' and 'other' will have the same m_drawPacketInFlight ptr with this change although before this->m_drawPacketInFlight was getting new-ed. So this->m_drawPacketInFlight was different from other.m_drawPacketInFlight

[the remark this answers is not in the document]
Comment (jhmueller-huawei) [reply Gems/Atom/RHI/Code/Source/RHI/DrawPacketBuilder.cpp:48]: Doesn't this call `operator=` below now which creates a new `m_drawPacketInFlight` just like this method did?

[the remark this answers is not in the document]
Comment (moudgils) [reply Gems/Atom/RHI/Code/Source/RHI/DrawPacketBuilder.cpp:48]: Ahh. Yes that works. I was on an older commit of the engine locally and didn't see an operator== in that class. Checked the latest version and it is there. Thanks for clarifying.

**Review by moudgils — APPROVED** [review]

**Comment by jhmueller-huawei** [comment]
> Do you have an accompanying atom-sampleviewer PR for these changes? To test multi-devices you can try the RPI -> MultiGPU sample there.

Any info on this?

**Review by jhmueller-huawei — APPROVED** [review]

**Comment by AMZN-Gene** [comment]
https://github.com/PopcornFX/O3DEPopcornFXPlugin/issues/80

**Pull request merged**